---
title: Analysis code used for the QSP NAFLD paper
output:
  pdf_document: default
  html_notebook: default
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

# TODO list
Create example data frames of data not included due to licensing 
Create README
Automate code testing
Consolidate my packages and add in any documentation


# Overview
The first section, [Analysis](#Analysis) has the code used in the computational analsyis of [@QSPpaper]. Intermediate data files are stored throughout the notebook as commented out `RDS` files. These are typically saved when creating the intermediate data is both/or time consuming or used at branch points within the analysis. Some of the code chunks rely on functions from my personal packages which are not on `CRAN` or `BioConductor` installed from github. 

The second section [Results](#Results) contain the code used to create the tables, figures, and data files for this paper.  

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

# Analysis 
## Installing my packages
```{r}

require(devtools)

# My package(s)
my_pkgs <- c("lefutils", "basicOmics")

if(!any(my_pkgs %in% row.names(installed.packages()))){
 devtools::install_github(paste0("lefeverde/", my_pkgs))
}

```
## Loading required packages
```{r}
require(tidyverse)
require(biomaRt)
require(matrixStats)
require(edgeR)
require(reshape2)
require(ggplot2)
require(lefutils)
require(basicOmics)

require(limma)
require(sva)
require(GSVA)
require(dendextend)

```



## Pre-processing pipeline
### Creating gene metadata
```{r}

#TODO ask DiStefano about including this data
gene_dir <- 'data_raw/DiStefano_raw_data/kallisto/'

# Getting ENS ids from 1st tsv
ens_trans <- 
  paste0(gene_dir, list.files(gene_dir)[[1]]) %>% 
  read_delim(., "\t") %>% 
  pull(target_id) %>% 
  str_split(., "\\.") %>% 
  map(1) %>% 
  unlist


attributes <- c("ensembl_gene_id", "ensembl_transcript_id")


ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl", version=94)

tx2gene <-  getBM(attributes, "ensembl_transcript_id", ens_trans, ensembl)
# saveRDS(tx2gene, "data_raw/tx2gene.rds")

attributes <- c("ensembl_gene_id",
                "external_gene_name",
                "gene_biotype",
                "description")

ens_genes <- unique(tx2gene$ensembl_gene_id)
gene_annots <- getBM(attributes, "ensembl_gene_id", ens_genes, ensembl)
#save(gene_annots, file = "data_raw/gene_annots.RData")

```

### Creating initial txi object
```{r}

tx2gene <- readRDS("data_raw/tx2gene.rds")
p <- readRDS('data_raw/DiStefano_raw_data/pdata.rds')
files <- 
  paste0(p$file_name, '.gz') %>% 
  paste0('data_raw/DiStefano_raw_data/kallisto/', .)
names(files) <- row.names(p)
txi <- tximport(files, type = 'kallisto', tx2gene = tx2gene, ignoreTxVersion = T, countsFromAbundance = 'lengthScaledTPM')

#saveRDS(list(p, txi), file='data_raw/DiStefano_txi_and_p_192.rds')
```


### Identifying mis-labeled samples 
```{r}


sex_linked_genes <- c('ENSG00000131002', 'ENSG00000183878', 'ENSG00000067048')

txi_n192 <- readRDS('data_raw/DiStefano_txi_and_p_192.rds')
p192 <- txi_n192[[1]]
txi_n192 <- txi_n192[[2]]
cnts <- txi_n192$counts
cnts_per_million <- cnts[rowMedians(cnts) > 1,] %>%
  data.frame(.) %>%
  cpm(., log = FALSE)

plot_data <- 
  data.frame(cnts_per_million[sex_linked_genes[1],], 
             sex=p192$sex, 
             sample_id=p192$dcl_patient_id) %>%
  melt(.) %>%
  dplyr::select( -one_of('variable')) %>%
  arrange(., value)

```


### Sex is incorrect for some samples 
Since this gene is Y-linked, I expect the males to have much higher expression than the females. I think the non-zero expression in the females can be attributed to multi-mapping reads. Kallisto does mapping a little differently, and so these are not discarded by default like with STAR and HISAT2.

```{r}
female_cutoff <-  plot_data %>%
  dplyr::filter(., sex =='Female') %>%
  arrange(., value) %>% 
  pull() %>% 
  quantile(., probs = .95) 

male_cutoff <-  plot_data %>%
  dplyr::filter(., sex =='Male') %>%
  arrange(., value) %>% 
  pull() %>% 
  quantile(., probs = .05)

suspicious_samples <- list(
  dplyr::filter(plot_data[plot_data$sex == 'Male',], value < male_cutoff),
  dplyr::filter(plot_data[plot_data$sex == 'Female',], value > female_cutoff) 
) %>% 
  lapply(., function(x){
    x$sample_id
  }) %>% 
  do.call(c, .)

```

### Creating new subsetted txi 
```{r}
suspicious_samples <- 
  c("DLDR_0189","DLDR_0031","DLDR_0022","DLDR_0017","DLDR_0131", "DLDR_0130","DLDR_0140","DLDR_0122","DLDR_0123","DLDR_0111")

load('data_raw/tx2gene.RData')
p192 <- readRDS('data_raw/DiStefano_raw_data/pdata.rds')
p <- p192 %>%
  dplyr::filter(., !dcl_patient_id %in% suspicious_samples)
row.names(p) <- p$dcl_patient_id

files <- 
  paste0(p$file_name, '.gz') %>% 
  paste0('data_raw/DiStefano_raw_data/kallisto/', .)
names(files) <- row.names(p)
txi <- tximport(files, type = 'kallisto', tx2gene = tx2gene, ignoreTxVersion = T, countsFromAbundance = 'lengthScaledTPM')
#saveRDS(list(p, txi), file='DiStefano_txi_and_p_182.rds')


```

### Creating cleaned expression matrix object and running SVA
```{r}

load('data_raw/gene_annots.RData')
txi_list <- readRDS('data_raw/DiStefano_txi_and_p_182.rds')
p <- txi_list[[1]]
txi <- txi_list[[2]]

mod <- basicOmics::better_model_matrix(~0 + diagnosis, data=p)
mod0 <- model.matrix(~1, data=p )
dge <- edgeR::DGEList(counts =  txi$counts, 
                      samples = p,
                      genes = gene_annots[row.names(txi$counts),])

dge <- dge[edgeR::filterByExpr(dge, group = dge$samples$diagnosis),]
dge <- dge[(rowSums(dge$counts > 1) >= .75*dim(dge)[2]),]
dge <- dge[!is.na(dge$genes$entrezgene),]

v <- voom(dge, mod, normalize.method = 'quantile')
svobj <- sva(v$E, mod=mod, mod0=mod0)
# saveRDS(list(v, svobj), 'data_raw/DiStefano_diagnosis_n182_voom_qn_and_svobj.rds')

```


## GSVA pathway cluster analysis
### Loading KEGG metadata
These files are not going to be included because of licensing issues with KEGG. The MSigDB pathways can be found at this [link](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/MSigDB_v7.0_Release_Notes).

```{r}

# This file wont be included due to licensing but can be obtained from here:
# https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/MSigDB_v7.0_Release_Notes
# The uncommented code has example data. 
#TODO Add code for creation of this tibble 
pth_data <- 
  # read_rds('../reference_files/msigdb_v7.0/msigdb_v7.0_all_tibble.rds') %>% 
  read_rds("data_raw/example_msigdb_v7.0_all_tibble.rds") 
  filter(SUB_CATEGORY_CODE == "CP:KEGG") %>% 
  dplyr::select(STANDARD_NAME,EXACT_SOURCE) %>% 
  setNames(., c('name','id')) 


# These pathways were causing issues and so they've been removed from the analysis
pths_to_remove <- c("KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM", "KEGG_FOLATE_BIOSYNTHESIS")
pth_data <- pth_data %>% filter(!name %in% pths_to_remove)


# This file was created by compiling the KEGG hiearchy files: "br08901.json","br08902.json", "br08904.json", and "br08907.json".
# These files can be found at the following link https://www.genome.jp/kegg-bin/get_htext?br08902 by using the search function 
# and downloading and parsing the jsons. The uncommented code has example data. 
hier_kegg <- 
  # read_rds('../thesisAnalysis/data_raw/kegg_pathway_hierarchy.rds') %>% 
  read_rds("data_raw/example_kegg_pathway_hierarchy.rds")
  left_join(pth_data, .) %>% 
  arrange(category_level_1)


# File can be directly downloaded from the link: https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/MSigDB_v7.0_Release_Notes
# 
cp_kegg <- 
  # '../reference_files/msigdb_v7.0/c2_kegg_entrez_msigdb_v7.0.gmt' %>% 
  "data_raw/example_c2_kegg_entrez_msigdb_v7.0.gmt" %>% 
  clusterProfiler::read.gmt(.) %>% 
  filter(term %in% hier_kegg$name) %>% 
  split(., .$term) %>% 
  lapply(., function(x){
    x$gene
  })


kegg_annots <-  
  hier_kegg$category_level_1 %>% set_names(., hier_kegg$name)


```

### Loading and pre-processing gene expression data 
Since I'm going run GSVA, I'm removing the batch effects predicted by SVA instead of including them in the downstream deferentially enriched pathway analysis. This is because the SVA matrix will be the wrong size. I'm also uniquefying the genes by variance. In other words, of the genes which match to the same entrez gene ID, I'm choosing the one that has the most variance.

```{r}
load('data_raw/gene_annots.RData') 

v <-  read_rds('data_raw/DiStefano_diagnosis_n182_voom_qn_and_svobj.rds')[[1]]
svobj <- read_rds('data_raw/DiStefano_diagnosis_n182_voom_qn_and_svobj.rds')[[2]]
p <-  v$targets

e_rm <- 
  removeBatchEffect(v, covariates = svobj$sv, design = v$design)
e_rm <- 
  uniquefy_by_variance(e_rm, gene_annots, 'entrezgene') 
row.names(e_rm) <- gene_annots[row.names(e_rm),]$entrezgene



v_e <- uniquefy_by_variance(v$E, gene_annots, 'entrezgene')
row.names(v_e) <- gene_annots[row.names(v_e),]$entrezgene


```

### Running GSVA and Clustering the output
I'm converting the gene x patient expression matrix into a pathway x patient enrichment matrix via [GSVA](https://www.bioconductor.org/packages/release/bioc/html/GSVA.html). I then perform a clustering analysis on the pathway x patient enrichment matrix. 

```{r}

gsva_res_sva <- 
  gsva(e_rm, cp_kegg, method='gsva',  min.sz=1, max.sz=100000)


tmp_data <- gsva_res_sva[names(kegg_annots),row.names(p)]

# Scaling row wise
tmp_data <-
  t(scale(t(tmp_data)))
tmp_data <- as.data.frame(tmp_data, check.names=FALSE)

col_clust <- 
  as.dist(1 - cor(tmp_data, method = 'pearson')) %>% 
  hclust(., "ward.D2")



```

#### Showing a dendrogram of the results
This chunk just creates a dendrogram of the clustering, like the one in figure 2. First, I create a color map of the different diagnoses, then the actual plot itself

```{r}

library(RColorBrewer)
colours <- brewer.pal(n = 7, name = "Paired")

# Ok so creating the color map was a bit involved. I started with 
# a palette of discrete colors, 2 for each group (e.g., Fibrosis)
# and then interpolated them so that I had darker colors for the 
# advanced grade/stage within each group.

groups <-  
  p[,c('group', 'diagnosis')] %>% 
  distinct() %>%
  dplyr::slice(-1) %>% 
  arrange(diagnosis)

col_map <- 
  tibble(idx=sort(rep(1:3, 2)),colours=colours[1:6]) %>% 
  split(., .$idx) %>% 
  lapply(., function(x){
    x$colours
  })

col_map2 <- 
  lapply(seq_along(col_map), function(i){
    x <- unique(groups$group)[i]
    tmp_lvls <- groups %>% 
      filter(group==x) %>% 
      pull(diagnosis) %>% 
      droplevels %>% 
      as.character
    
    out_cols <- 
      colorRampPalette(col_map[[i]])(length(tmp_lvls)) %>% 
      setNames(., tmp_lvls)
  }) %>% do.call(c, .)

col_map <- 
  c(NORMAL=colours[length(colours)], col_map2) 

# Getting the diagnosis for the labels in the dendrogram 
ordered_colors <- p$diagnosis[match(labels(col_clust), row.names(p))]
ordered_colors <- col_map[ordered_colors]

dend <- as.dendrogram(col_clust)
labels_colors(dend) <- ordered_colors

# Finally, the plot
par(cex=.5)
plot.new()
plot(dend)

# save(dend, col_map, tmp_data,file =  "data_for_figure2_heatmap.RData")

```


### Creating new groups based on clustering
I'm cutting the dendogram at the 3rd level, since 3 distinct clusters naturally form (see figure 2 and previous section). This is used to create new groupings of the samples (see methods section of paper).

Apparently the default `cutree` will sort the clusters by number of items in each cluster. This is different than `ComplexHeatmap`. The `dendextend` version has an option to turn this "feature" off. To do this, set the confusingly named argument `order_clusters_as_data` to `FALSE`. 

```{r}

coefs_from_clusters <- 
  dendextend::cutree(col_clust, 3, order_clusters_as_data=FALSE) %>% 
  enframe %>% 
  as.data.frame %>% 
  setNames(., c('sample_id', "cluster_idx"))
coefs_from_clusters$cluster_idx <- paste0('c', coefs_from_clusters$cluster_idx)
coefs_from_clusters$cluster_idx <- 
  factor(coefs_from_clusters$cluster_idx, levels = unique(coefs_from_clusters$cluster_idx))
row.names(coefs_from_clusters) <- coefs_from_clusters$sample_id
coefs_from_clusters$sample_id <- NULL
lvls <- levels(coefs_from_clusters$cluster_idx)

```

### Running LIMMA-VOOM using the new groupings
#### Identifying differentially enriched pathways 
```{r}

group_cont <- c("c2-c1", "c3-c1", "c3-c2")

mod <- 
  better_model_matrix(~0 + cluster_idx, data = coefs_from_clusters)

cont_mod <- 
  makeContrasts(contrasts = group_cont, levels=data.frame(mod, check.names = FALSE))

fit <-  t(scale(t(gsva_res_sva))) %>% # Row scaling
  lmFit(., data.frame(mod)) %>% 
  contrasts.fit(., cont_mod) %>% 
  eBayes


res <- get_limma_results(fit, group_cont)
names(res)[2] <- 'name'
res <- right_join( hier_kegg, res)

# saveRDS(res, 'lv_on_gsva_msig_kegg_pathways_from_row_scaled_clusters.rds')

```

#### Identifying differentially expressed genes
```{r}
fit <-   
  lmFit(v, data.frame(mod)) %>% 
  contrasts.fit(., cont_mod) %>% 
  eBayes

res <- get_limma_results(fit, group_cont)
# saveRDS(res, "gene_res_from_clusters.rds")

```

## Creating gene signatures 
### Loading and prepping the requisite data

```{r}
cat_pths <- 
  read_delim("data_raw/NAFLD_progression2path_03-09-2021.txt", 
             "\t",
             col_names = FALSE) %>% 
  set_names(., c("nafld_category", "id"))

kegg_db <- 
  read_rds("data_raw/c2_kegg_entrez_msigdb_v7.0_database.rds") %>% 
  separate_rows(entrezgene, sep = '\\|') %>% 
  select(name, id, entrezgene)

sig_pth_res <- 
  read_rds("data_raw/lv_on_gsva_msig_kegg_pathways_from_row_scaled_clusters.rds") %>% 
  filter(adj.P.Val < .001)

cols_to_keep <- c("coefficient","name","id", "logFC", "adj.P.Val" )
sig_pth_res <- sig_pth_res[,cols_to_keep]

names(sig_pth_res)[4:5] <- paste0("pathway_", names(sig_pth_res)[4:5])

pths_to_keep <- cat_pths %>% filter(nafld_category %in% 1:4) %>% pull(id)
sig_pth_res <- sig_pth_res %>% filter(id %in% pths_to_keep)
sig_pth_res <- inner_join(sig_pth_res, cat_pths)
sig_pth_res <- inner_join(sig_pth_res, kegg_db)

gene_res <- read_rds('data_raw/gene_res_from_clusters.rds') %>% 
  # filter(adj.P.Val < .05)
  filter(adj.P.Val < .001)
gene_res$entrezgene <- as.character(gene_res$entrezgene)

cols_to_keep <- c("coefficient","external_gene_name", "entrezgene", "logFC", "adj.P.Val")
gene_res <- gene_res[,cols_to_keep]
names(gene_res)[4:5] <- paste0("gene_", names(gene_res)[4:5])

gene_pth_filt <- inner_join(gene_res, sig_pth_res)


```


### Creating the gene sigantures 
```{r}

upreg_genes <- gene_pth_filt %>% 
  filter(gene_logFC > 0) %>% 
  group_by(coefficient, nafld_category) %>% 
  summarise(up_genes_entrez=paste(unique(entrezgene), collapse = ";"),
            n_upreg=length(unique(entrezgene)))
  

downreg_genes <- gene_pth_filt %>% 
  filter(gene_logFC < 0) %>% 
  group_by(coefficient, nafld_category) %>% 
  summarise(down_genes_entrez=paste(unique(entrezgene), collapse = ";"),
            n_downreg=length(unique(entrezgene)))

unweighted_gene_sig <- 
  left_join(upreg_genes, downreg_genes) %>% 
  arrange(coefficient, nafld_category) %>% 
  ungroup
unweighted_gene_sig$sig_idx <- 1:12
unweighted_gene_sig <- unweighted_gene_sig %>%  relocate(sig_idx)

# save(unweighted_gene_sig, file = "gene_signatures_03-18-21.RData")

```

## Creating database DrugBank LINCS 
```{r}

cols_to_keep <- 
  c("pert_id",
    "pert_iname", 
    "pert_type",
    "is_touchstone", 
    "inchi_key_prefix", 
    "inchi_key", 
    "canonical_smiles", 
    "pubchem_cid")

# This file was created by joining the metadata files listed here: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742
#TODO figure out if this can be included or not 
tmp_map <- read_rds('data_raw/GSE92742_inst_metadata.rds') %>% 
  filter(pert_type == "trt_cp") %>% 
  distinct
tmp_map <- tmp_map[,cols_to_keep]

# This is example of the data. The full DrugBank database  can be downloaded from here:
# https://go.drugbank.com/releases/5-1-4/release_notes
drugbank_data <- read_rds("data_raw/example_drugbank_v5.1.4_data.rds")
  separate_rows(., Synonyms, sep = '\\|')
names(drugbank_data) <- str_to_lower(names(drugbank_data))

dd1 <- drugbank_data[,-3]
names(dd1)[2] <- 'drug_name'
dd2 <- drugbank_data[,-2]
names(dd2)[2] <- 'drug_name'
tmp_db_data <- 
  rbind(dd1, dd2) %>% 
  distinct 


names(tmp_db_data)[which(names(tmp_db_data) == 'smiles')] <- 'canonical_smiles'
names(tmp_db_data)[which(names(tmp_db_data) == 'pubchem_compound_id')] <- 'pubchem_cid'
tmp_db_data$pert_iname <- str_to_lower(tmp_db_data$drug_name)

cols_to_keep <- c('canonical_smiles', 'pubchem_cid', 'pert_iname')
db_cmap_matches <- 
  lapply(cols_to_keep, function(x){
    db_sub <- 
      tmp_db_data[,c('drugbank_id', x)] %>% 
      na.omit %>% 
      distinct
    inner_join(tmp_map, db_sub, na_matches = "never")
}) %>% do.call(rbind, .)

# Removing extra columns and then re-adding based on trt_cps file
db_cmap_matches <- db_cmap_matches %>% 
  select(pert_id, pert_iname, drugbank_id) %>% 
  left_join(., drugbank_data[,c("drugbank_id", "name")]) %>% 
  distinct
man_annots <- read_rds("data_raw/manual_lincs_drugbank_drug_annots.rds") # these were matches done manually 
db_cmap_matches <- db_cmap_matches %>% filter(!pert_id %in% man_annots$pert_id)
db_cmap_matches <- rbind(db_cmap_matches, man_annots)

dbids_to_remove <- c("DB07159", "DB02507", "DB02731", "DB00452", "DB00624") # These are duplicates with extraneous DB IDs

db_cmap_matches <- db_cmap_matches %>% filter(!drugbank_id %in% dbids_to_remove)

# saveRDS(db_cmap_matches, file = 'GSE92742_perts_in_DrugBank_03-30-21.rds')

```

## Performing the CMap analysis 

```{r}

# Figuring out which perts to analyze
meta_cols_to_keep <- 
  c("sig_id",
    "pert_id",
    "pert_iname")

trt_cp <-
  read_rds('data_raw/GSE92742_inst_metadata.rds') %>% 
  filter(pert_type == "trt_cp") %>% 
  distinct

trt_cp <- trt_cp[,meta_cols_to_keep]

#TODO think refactoring such that this works with updated file 
# right now it doesn't for the end result, it's just annoying have
# what amounts to duplicate files. 
drgbnk_db <- 
  readRDS('data_raw/GSE92742_perts_in_DrugBank_03-30-21.rds') %>% 
  dplyr::select(pert_id, pert_iname, drugbank_id) %>% 
  na.omit %>% 
  distinct

trt_cp <- inner_join(trt_cp, drgbnk_db)
# saveRDS(trt_cp, "sig_id_drugbank_cmap_overlap.rds")

load("data_raw/gene_signatures_03-18-21.RData")

# This file was created by downloading the CMap contest data csv. It can be found at the clue.io data library under contests. 
# The contest is entitled "CMap QUery Speedup Challenge" and the file is "sig_score_n476251x10174.csv.gz". It was then converted
# into a gctx file to make reading it in chunks easier. 
#TODO Add code for creating this file
gctx_file <- '/zfs1/dtaylor/cmap_data/clue_test_data/sig_score_trt_cp_n204975x10174.gctx'
cids <- read_gctx_meta(gctx_file,dim = "col") %>% pull
gids <- read_gctx_meta(gctx_file,dim = "row") %>% pull

# Filters all but sigs with DrugBank pert
cids <- cids[cids %in% trt_cp$sig_id]

cid_chunks <- chunker(cids, 6)



gene_sets <- unweighted_gene_sig %>% 
  select(sig_idx, up_genes_entrez, down_genes_entrez)


unweighted_res_sub <-
  map(cid_chunks, function(y){
    sig_list <- 
      gctx_to_sigList(gctx_file, y, decreasing_order = FALSE)
    out_res <- multi_broad_wrapper(gene_sets, sig_list, 7)
  }) %>% 
  bind_rows

unweighted_random_res_sub <- 
  map(1:nrow(gene_sets), function(i){
    x <- gene_sets[i,]
    up_genes <- str_split(x[,2], pattern = ';') %>% unlist
    up_genes <- up_genes[up_genes %in% gids]
    down_genes <- str_split(x[,3], pattern = ';') %>% unlist
    down_genes <- down_genes[down_genes %in% gids]
    random_cs <- 
      get_random_cmap_scores(up_genes, down_genes, gctx_file, cids, n_perm=5e+05, num_chunks = 6, n_cores = 12)
  })


# save(unweighted_res_sub, unweighted_random_res_sub, file = "unweighted_cmap_res_db_sub_03-22-21.RData")


```

### Calculating the p-values 
```{r}

# File not currently included in the data_raw dir because it's 10g. 
#TODO make shrink and include this file 
load("data_raw/unweighted_cmap_res_db_sub_03-22-21.RData")
# Removing extra columns to save on mem
unweighted_random_res_sub <- unweighted_random_res_sub %>% 
  map(function(x){
    x$r_up_genes <- NULL
    x$r_down_genes <- NULL
    return(x)
  })

unweighted_res_sub$up_genes <- NULL
unweighted_res_sub$down_genes <- NULL
gc()

plan(multisession, workers=15)
unweighted_res_sub <- split(unweighted_res_sub, unweighted_res_sub$sig_idx)
unweighted_res_w_p_value <- add_p_vals_to_cmap_para(unweighted_res_sub, unweighted_random_res_sub)
# saveRDS(unweighted_res_w_p_value, "unweighted_res_sub_w_p-values_03-22-21.rds")

```

### Processing raw CMap results
```{r}

meta_cols_to_keep <- 
  c("sig_id",
    "pert_id",
    "pert_iname")

trt_cp <-
  read_rds('data_raw/GSE92742_inst_metadata.rds') %>% 
  filter(pert_type == "trt_cp") %>% 
  distinct

trt_cp <- trt_cp[,meta_cols_to_keep]

load("data_raw/gene_signatures_03-18-21.RData")


cmap_res <- read_rds("data_raw/unweighted_res_sub_w_p-values_03-22-21.rds")
# Fixing some issues with p-values, since I didn't divide by 
# length. 
#TODO fix this so the code isn't neccessary 
cmap_res <-  cmap_res %>% group_by(sig_idx) %>% mutate(p_value=p_value/max(p_value))
cmap_res <-  cmap_res %>% group_by(sig_idx) %>% mutate(fdr_p_value=p.adjust(p_value, method = 'fdr'))


sig_idx_map <- read_csv("data_raw/Data_file_S3_Gene_signatures.csv") %>% 
  select(1:3)
names(sig_idx_map)[1] <- "sig_idx"

sig_cmap <- inner_join(trt_cp, cmap_res) %>% 
  group_by(sig_idx, pert_iname) %>% 
  arrange(cmap_score, .by_group=TRUE) %>% 
  slice(1) %>% 
  ungroup
sig_cmap <- left_join(sig_cmap, sig_idx_map)

drg_db <- read_rds("data_raw/GSE92742_perts_in_DrugBank_03-30-21.rds") 
sig_cmap <- left_join(sig_cmap, drg_db)

sig_cmap <- sig_cmap %>% 
  group_by(sig_idx) %>% 
  arrange(cmap_score, .by_group=TRUE) %>% 
  mutate(drug_idx=row_number()) %>% 
  ungroup

fdr05_tbl <- sig_cmap %>% 
  filter(fdr_p_value < .05) %>% 
  # select(-drugbank_id, -name) %>% 
  distinct %>% 
  group_by(sig_idx) %>% 
  arrange(cmap_score, .by_group=TRUE) %>% 
  slice(1:20) %>% 
  ungroup
# saveRDS(fdr05_tbl, "fdr05_tbl_top20_drugs.rds")


```

## Network proximity 
The python code for this portion of the analysis can be downloaded from this [repo](https://github.com/emreg00/toolbox). I ran everything in a local directory with the package. 

### Creating a file of the drug targets
```{r}


drug_list <- read_rds("data_raw/manual_lincs_drugbank_drug_annots.rds")
perts <- read_rds("data_raw/GSE92742_perts_in_DrugBank_03-30-21.rds")
drugbank_data <- 
  read_rds("data_raw/drugbank_db_metadata_v5.1.4.rds") %>% 
  select(drugbank_id, targets) %>% 
  separate_rows(targets, sep ="\\|") 
  
names(drugbank_data)[2] <- "gene_name"
drugbank_data$gene_name <- gsub('\\-', '', drugbank_data$gene_name)

xx <- as.data.frame(org.Hs.eg.db::org.Hs.egALIAS2EG) %>%
  setNames(., c("entrezgene", "gene_name"))
drugbank_data <- left_join(drugbank_data, xx) %>% na.omit
drug_list <- left_join(drug_list, drugbank_data)

as_tgts <-  paste(c("5599", "5602", "4293", "5601"), collapse = ';' )
as_tgts <- tibble(drug_name= "AS-601245", targets=as_tgts)

drug_list$fixed_drug_names <- drug_list$pert_iname
drug_list$fixed_drug_names[str_detect(drug_list$pert_iname, "BR")] <- drug_list$name[str_detect(drug_list$pert_iname, "BR")]
drug_list <- drug_list %>% filter(drugbank_id !="DB08022")

out_list <- split(drug_list, drug_list$fixed_drug_names) %>% 
    map(function(x){
        tmp_genes <- unique(x$entrezgene)
        enframe(paste(tmp_genes, collapse = ';'),value = 'targets')
    }) %>% rbind_named_df_list(., "drug_name")

out_list$name <- NULL
out_list <- rbind(out_list, as_tgts)

# write_csv(out_list, "targets_from_top_cmap_drugs_03-27-21.csv")
```

### Calculating network proximity 
Again the python code can be found [here](https://github.com/emreg00/toolbox). I excuted in this in the top-level dir for this package.
```{python}
import random
import toolbox
from toolbox import wrappers

# Loading the data 
drug_targets = [i.strip().split(',') for i in open("targets_from_top_cmap_drugs_03-27-21.csv")] 
drug_targets = drug_targets[1:]
network = wrappers.get_network("biosnap_liver_ppi_network.txt", only_lcc = True)
degs = [i.strip() for i in open("updated_degs_in_network.txt")]

# Running analysis and writing results 
with open("network_proximity_results_03-27-21.csv", "w+") as f: 
    hd = "db_id,d,z,mean,sd\n"
    f.write(hd)
    for i in drug_targets:
        db_id = i[0]
        targets = i[1].split(';')
        print(i)
        try:
            d, z, tmp_val = wrappers.calculate_proximity(network, targets, degs,  min_bin_size=90)
            out_line = [db_id, d, z, tmp_val[0], tmp_val[1]]
            out_line = [str(i) for i in out_line]
            out_line = ','.join(out_line)
            f.write(out_line + '\n')
        except:
            continue

```




# Results 
## Figures 
### Figure 2: Heatmap 

```{r}

library(ComplexHeatmap)

load("data_raw/data_for_figure2_heatmap.RData")


col_tits <- 
  c("Normal &\nSteatosis\n(N&S)",
    "Predominately\nLobular\nInflammation\n(PLI)",
    "Predominately\nFibrosis\n(PF)")
cluster_colors <- c("#6a3d9a","#ff7f00", "#b15928")

v <-  read_rds('data_raw/DiStefano_diagnosis_n182_voom_qn_and_svobj.rds')[[1]]
p <-  v$targets

cols_to_keep <- c("diagnosis", "bmi_surg", "sex", "age","icd_250")

annot_map <- p[,cols_to_keep] %>% 
  rownames_to_column %>% 
  map_if(is.factor, as.character) %>% 
  bind_cols %>% 
  as.data.frame
row.names(annot_map) <- annot_map$rowname
annot_map <- annot_map[,-1]
colnames(annot_map) <- str_to_title(colnames(annot_map))

colnames(annot_map)[ncol(annot_map)] <- "T2D"



sex <- c('palevioletred1', 'royalblue') %>% 
  setNames(., c('Female', 'Male'))

col_map <- 
  list(Sex=sex, Diagnosis=col_map)

age_plot <- anno_points(annot_map['Age'], 
                        size = unit(1.5, "mm"),
                        axis_param =list(side="right",
                                         at=c(30,  60),
                                         gp=gpar(fontsize = 4.5,
                                                 fontface='bold')))


bmi <- anno_barplot(annot_map['Bmi_surg'],
                    axis_param = list(side="right",
                                      at=c(30, 90 ),
                                      gp=gpar(fontsize = 4.5, fontface='bold')))


tmp_ht <-
  Heatmap(tmp_data, 
          # Column args
          show_column_names = FALSE, 
          cluster_columns= col_dend,
          column_title = col_tits,
          column_title_gp = gpar(fontsize=12, fontface="bold", col=cluster_colors),
          # column_dend_reorder = TRUE,
          column_split = 3,
          column_gap = unit(2, "mm"),
          # Row args 
          cluster_rows = TRUE,
          cluster_row_slices = FALSE,
          show_row_names = FALSE,
          show_row_dend = FALSE,
          row_split = kegg_annots,
          row_title_gp = gpar(fontsize=12, fontface="bold"),
          row_title_rot = 0,
          # Annotations

          top_annotation  = ha,
          # Dimensions
          height = unit(.025*nrow(tmp_data), 'inch'),
          width = unit(.025*ncol(tmp_data), 'inch'),
          # Misc parameters
          show_heatmap_legend = TRUE,
          heatmap_legend_param =
            list(title = "ES",
                 just = "right",
                 legend_width = unit(.0125*ncol(tmp_data), 'inch'), 
                 grid_width =unit(.75, 'cm'),
                 direction = "horizontal",
                 title_position = "lefttop")
          
  )

ht_height <- 
    ComplexHeatmap:::height(draw(tmp_ht, heatmap_legend_side="bottom")) %>% 
    convertUnit(., 'inch') %>% 
    ceiling
ht_width <- 
    ComplexHeatmap:::width(draw(tmp_ht, heatmap_legend_side="bottom")) %>% 
    convertUnit(., 'inch') %>% 
    ceiling



pdf(file = "figure2_heatmap.pdf",
    height = ht_height,
    width = ht_width,
    bg = "transparent")
# draw(tmp_ht, heatmap_legend_side="bottom")
draw(tmp_ht, show_annotation_legend = FALSE, show_heatmap_legend=FALSE)
dev.off()



```

### Figure 3: Pathway summaries 
```{r}
sig_pth_res <- 
  read_rds("data_raw/lv_on_gsva_msig_kegg_pathways_from_row_scaled_clusters.rds") %>% 
  filter(adj.P.Val < .001)
coef_nms <- c("PLI vs. N&S", "PF vs. N&S", "PF vs. PLI" ) %>% 
  setNames(.,unique(sig_pth_res$coefficient))

sig_pth_res$coefficient <- factor(coef_nms[sig_pth_res$coefficient], levels = coef_nms)

cat_pths <- 
  read_delim("data_raw/NAFLD_progression2path_03-09-2021.txt", 
             "\t",
             col_names = FALSE) %>% 
  set_names(., c("disease_category", "id"))

cat_pths$disease_category <- paste0("C", cat_pths$disease_category)

sig_pth_res <- left_join(sig_pth_res, cat_pths)

```

#### Top panel
```{r}
library(scales)
library(cowplot)

plot_data <- sig_pth_res %>% 
  select(coefficient, name, category_level_1) %>%
  distinct %>%
  group_by(coefficient, category_level_1) %>% 
  summarise(cnt=n()) %>% 
  mutate(prct=cnt/sum(cnt)) %>% 
  ungroup %>% 
  mutate(category_categorization="KEGG Groups")
names(plot_data)[2] <- "pathway_category"


plot_data2 <- sig_pth_res %>% 
  select(coefficient, name, disease_category) %>% 
  distinct %>% 
  group_by(coefficient, disease_category) %>% 
  summarise(cnt=n()) %>% 
  mutate(prct=cnt/sum(cnt)) %>% 
  ungroup %>% 
  mutate(category_categorization="NAFLD Categories")
names(plot_data2)[2] <- "pathway_category"



plot_data <- rbind(plot_data, plot_data2)
rm(plot_data2)
plot_data$category_categorization <- 
  factor(plot_data$category_categorization, levels = c("KEGG Groups", "NAFLD Categories"))
lvls <- c(unique(sig_pth_res$category_level_1), unique(cat_pths$disease_category)) %>% rev
plot_data$pathway_category <- factor(plot_data$pathway_category, levels=lvls)


plot_data$prct <- label_percent(accuracy = .1)(plot_data$prct)

plt <-  ggplot(data=plot_data, aes(x=pathway_category, y=cnt)) + 
  geom_bar(stat='identity',aes(fill=category_categorization))  + 
  theme(panel.border = element_rect(colour = "black", fill=NA, size=1),
        text=element_text(face="bold"),
        axis.title.y = element_blank(), 
        axis.title.x = element_text(face="bold", size=12, colour="black"),
        axis.text = element_text(face="bold", size=12, colour="black"),
        aspect.ratio = 1,
        strip.text = element_text(size =12),
        panel.spacing = unit(1.5, "lines"),
        legend.position = "none") + 
  facet_grid(category_categorization ~ coefficient    , scales = "free_y") +
  scale_y_continuous(expand = expansion(mult = c(0, .4))) +
  coord_flip()

plt <- plt + geom_text(aes(x=pathway_category, y=cnt, label=prct),size=3.5, hjust = -.075) + labs(y="Number of differentially enriched pathways")
ggsave("updated_fig3_pathway_summary.pdf",plt, height = 4.5, width = 8.75)

```

#### Top 10 pathway bar charts
```{r}
library(readr)
library(cowplot)



cats_to_keep <- c("C1", "C2", "C3", "C4")

pth_data <- read_csv("Data_file_S2_Differentially_enriched_pathways.csv") %>% 
  separate_rows(nafld_categories, sep = '\\|')


plot_data <- pth_data %>% 
  # select(-2) %>% 
  mutate(pathway_neg_log10_pval=-log(adj.P.Val)) %>% 
  distinct
  

plot_data <- plot_data %>%
  group_by(comparison) %>% 
  arrange(desc(pathway_neg_log10_pval), .by_group=FALSE) %>% 
  slice(1:10) %>% 
  ungroup

plot_data$comparison <- factor(plot_data$comparison, 
                               levels = c("PLI vs. N&S", "PF vs. N&S","PF vs. PLI" ))

# plot_data <- plot_data %>% arrange(comparison, desc(pathway_logFC))
#plot_data <- plot_data %>% arrange(comparison, pathway_logFC)
plot_data <- plot_data %>% arrange(comparison, pathway_neg_log10_pval)

plot_data$tmp_pth_name <- 
  paste(plot_data$comparison, plot_data$pathway_name, sep="_") %>% 
  factor(., levels = .)
plot_data$fill_factor <- factor(if_else(plot_data$pathway_logFC < 0, 1, 2))

plt <-  ggplot(data=plot_data, aes(x=tmp_pth_name, 
                                   y=pathway_neg_log10_pval,
                                  # alpha=pathway_neg_log10_pval,
                                   fill=pathway_logFC)) + 
  geom_bar(stat="identity") + 
  coord_flip() +
  # scale_x_discrete(labels=plot_data$pathway_name) + 
  facet_wrap(~comparison, nrow=3, scales = "free_y") +
  scale_x_discrete(label=function(x){
    x %>% 
      str_split(., '_') %>% 
      map_chr(2)
  }) + scale_fill_gradient2(low = "blue",mid = "white", high = "red")


plt$labels$fill <-  ""
plt <- plt  + theme(legend.title = element_text(face="bold"), legend.text = element_text(face="bold"))
ggsave("log2fc_legend_for_top10_pathways.pdf", get_legend(plt), width = .75, height = 1.5)


plt <- plt + 
  #labs(x = '', y="Pathway log2 fold change") + 
  guides(fill=FALSE) + 
  theme_cowplot() +
  theme(legend.position = "none",
        plot.margin = unit(c(0, 0, 0, 0), "cm"),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(), 
        text=element_text(face="bold")) 

ggsave("top_pathways_bar_plot.pdf", aligned_plots[[1]], height = 6.5, width = 7)

```

#### Creating a grid with the categories
```{r}
library(settleR)

cat_list <- pth_data %>% 
  select(disease_category, pathway_name) %>% 
  distinct %>% 
  filter(disease_category %in% cats_to_keep)

cat_list$disease_category <- 
  names(cats_to_keep)[match(cat_list$disease_category, cats_to_keep)]

# cat_list <- split(cat_list, cat_list$disease_category) %>% 
#   map(function(x) unique(x$pathway_name))

cat_list <- split(cat_list, cat_list$pathway_name) %>% 
  map(function(x) unique(x$disease_category))

cat_mat <- sets_to_matrix(cat_list) 
cat_mat[cat_mat == 0] <- NA
cat_mat <- melt(cat_mat) %>% na.omit
cat_mat <- cat_mat[,-3]
names(cat_mat) <- c("disease_category", "pathway_name")

plot_data <- left_join(plot_data, cat_mat)

grid_data <-  plot_data %>% 
    split(., .$tmp_pth_name) %>% 
    map(function(x) as.character(x$disease_category)) %>% 
    sets_to_matrix(.) 
grid_data <- grid_data[sort(row.names(grid_data)),] %>% melt %>% 
    setNames(., c("intersect_id","set_names","observed"))
grid_data$observed <- as.logical(grid_data$observed) 

grid_data$comparison <- str_split(grid_data$set_names, '_') %>% map_chr(1) %>% 
  factor(., levels = levels(plot_data$comparison))




v_max <- max(1, length(unique(grid_data$intersect_id)) -1)
v_breaks <- seq(1, v_max, by = 1)+ .5
h_max <- max(1, length(unique(grid_data$set_names)) - 1)
h_breaks <- seq(1, h_max, by = 1)+ .5

dot_size <- 4.25

p <-  ggplot(data=grid_data, aes(x=intersect_id, y=set_names)) +
  geom_point(size=dot_size, alpha=.25) +
  theme_empty()
# just plots empty dots

# p <- p +
#   geom_line(data=grid_data[grid_data$observed,], # adds the lines
#             aes(x=intersect_id, y=set_names, group=set_names),
#             size=dot_size/3,
#             inherit.aes = FALSE)
p <- p +
  geom_point(data=grid_data[grid_data$observed,], # Add `filled in points`
             aes(x=intersect_id, y=set_names),
             size=dot_size,
             inherit.aes = FALSE)
p <- p + theme(axis.text.y = element_blank(),
               plot.margin = unit(c(0, 0, 0, 0), "cm")) +
  geom_vline(xintercept = v_breaks) +
  geom_hline(yintercept = h_breaks)

p <-  p + facet_wrap(~comparison, nrow=3, scales = "free_y")

aligned_plots <- cowplot::align_plots(plt, p, align="hv", axis="tblr")


ggsave("disease_category_grid_plot.pdf", aligned_plots[[2]], height = 6.5, width = 6)


```
## Tables 
### Table S2
```{r}
library(flextable)
library(officer)
#TODO Refactor so this all uses the data files 
sig_pth_res <- 
  read_rds("data_raw/lv_on_gsva_msig_kegg_pathways_from_row_scaled_clusters.rds") %>% 
  filter(adj.P.Val < .001)
coef_nms <- c("PLI vs. N&S", "PF vs. N&S", "PF vs. PLI" ) %>% 
  setNames(.,unique(sig_pth_res$coefficient))

sig_pth_res$coefficient <- factor(coef_nms[sig_pth_res$coefficient], levels = coef_nms)

#TODO Refactor this bit 
cat_pths <- 
  read_delim("data_raw/NAFLD_progression2path_03-09-2021.txt", "\t", col_names = FALSE) %>% 
  set_names(., c("nafld_category", "id"))

#TODO Refactor this bit 
pth_map <- 
  read_delim("data_raw/fens_categories.txt", ";", col_names = FALSE) %>% 
  setNames(., c("nafld_category", "disease_category"))

cat_pths <- left_join(cat_pths, pth_map)
cat_pths$nafld_category <- paste0("C", cat_pths$nafld_category)
# sig_pth_res <- left_join(sig_pth_res, cat_pths)

nafld_category <- split(cat_pths, cat_pths$id) %>% 
  map(function(x){
    tmp_cats <-  x %>% 
      pull(nafld_category) %>% 
      paste0(., collapse = '|') %>% 
      enframe
  }) %>% rbind_named_df_list(., "id") %>% 
  select(-name)
names(nafld_category)[2] <- "nafld_categories"

sig_pth_res <- left_join(sig_pth_res, nafld_category)

#TODO File needs to be included somehow
pth_refs <- 
  read_excel("data_raw/pth_refs.xlsx", sheet = 1)
pth_refs <- pth_refs[,c(2,3)]

pth_refs <-  split(pth_refs, pth_refs$pathway_id) %>% 
  map(function(x){
    tmp_cats <-  x %>% 
      pull(pmids) %>% 
      paste0(., collapse = '|') %>% 
      enframe
  }) %>% rbind_named_df_list(., "id") %>% 
  select(-name)

names(pth_refs)[2] <- "PMID"


tbl_data <- left_join(sig_pth_res, pth_refs)
tbl_data <- left_join(tbl_data, nafld_category)
fixed_pathway_name <- paste0("(", tbl_data$id, ")") %>% 
  paste(tbl_data$description, .)
tbl_data$fixed_pathway_name <- fixed_pathway_name


cols_to_keep <- 
  c("coefficient", "fixed_pathway_name", "category_level_1","category_level_2","nafld_categories", "logFC", "adj.P.Val", "PMID")

tbl_data <- tbl_data[,cols_to_keep]


fixed_nms <- c("coefficient","Pathway name", "KEGG pathway group", "KEGG pathway subgroup", "NAFLD category", "log2 Fold change", "FDR corrected p-value","PMIDs")

split_tbls <- split(tbl_data, tbl_data$coefficient) %>% 
  map(function(x){
    tmp_tbl <- x %>% 
      arrange(logFC)
    names(tmp_tbl) <- fixed_nms
    return(tmp_tbl)
  })



```

## Data files 
### Data file S1
```{r}

hier_kegg <- 
  read_rds('data_raw/kegg_pathway_hierarchy.rds')

#TODO refactor here 
cat_pths <- 
  read_delim("data_raw/NAFLD_progression2path_03-09-2021.txt", 
             "\t",
             col_names = FALSE) %>% 
  set_names(., c("nafld_category", "id"))

pth_map <- 
  read_delim("data_raw/fens_categories.txt", ";", col_names = FALSE) %>% 
  setNames(., c("nafld_category", "disease_category"))

cat_pths <- left_join(cat_pths, pth_map)
cat_pths$nafld_category <- NULL


kegg_df <- 
  read_rds("data_raw/c2_kegg_entrez_msigdb_v7.0_database.rds") %>% 
  right_join(., cat_pths) %>% 
  separate_rows(., entrezgene, sep = '\\|')

per_gene_kegg <- split(kegg_df, kegg_df$entrezgene) %>% 
  map(function(x){
    kegg_pathway_names <- paste0(unique(x$description), collapse = '|')
    kegg_pathway_ids <- paste0(unique(x$id), collapse = '|')
    entrezgene <- unique(x$entrezgene)
    out_df <- tibble(entrezgene, kegg_pathway_names, kegg_pathway_ids)
  }) %>% do.call(rbind, .)

per_gene_kegg <- distinct(per_gene_kegg)


adj_p_val_thresh <- .05

cols_to_remove <- 
  c("gene_biotype","description")
gene_res <- read_rds('gene_res_from_clusters.rds') %>% select(-cols_to_remove)


gene_res$entrezgene <- as.character(gene_res$entrezgene)

sig_gene_res <- gene_res %>% 
  filter(adj.P.Val < adj_p_val_thresh) 

sig_gene_res <- left_join(sig_gene_res, per_gene_kegg)

names(sig_gene_res)[1:4] <- c("comparison", "ensembl_id", "gene_symbol", "entrez_gene_id")

clst_map <- 
  c("PLI vs. N&S", "PF vs. N&S", "PF vs. PLI") %>% 
  setNames(., unique(sig_gene_res$comparison))
sig_gene_res$comparison <- unlist(clst_map[sig_gene_res$comparison])
sig_gene_res <- sig_gene_res %>% 
  relocate(gene_symbol, .before=ensembl_id) %>% 
  relocate(entrez_gene_id, .before=ensembl_id)

# write_csv(sig_gene_res, file = "Data_file_S1_DEG_details.csv")

```

### Data file S2
```{r}

cols_to_remove <- c("name", "category_id")

sig_pth_res <- 
  read_rds("data_raw/lv_on_gsva_msig_kegg_pathways_from_row_scaled_clusters.rds") %>% 
  filter(adj.P.Val < .001) %>% 
  select(-cols_to_remove)

sig_pth_res <- sig_pth_res %>% relocate(description, .before=id)
names(sig_pth_res)[1:2] <- c("comparison", "pathway_name")

sig_pth_res$comparison <- unlist(clst_map[sig_pth_res$comparison])
names(sig_pth_res)[4:5] <-  c("kegg_pathway_group", "kegg_pathway_subgroup")


nafld_category <- 
  read_delim("data_raw/NAFLD_progression2path_03-09-2021.txt", 
             "\t",
             col_names = FALSE) %>% 
  set_names(., c("nafld_categories", "id"))

nafld_category$nafld_categories <- paste0("C", nafld_category$nafld_categories)
nafld_category <- split(nafld_category, nafld_category$id) %>% 
  map(function(x){
    ncats <- paste(unique(x$nafld_categories), collapse = "|")
    id <- unique(x$id)
    tibble(nafld_categories=ncats, id)
  }) %>% bind_rows
sig_pth_res <- left_join(sig_pth_res, nafld_category)

# write_csv(sig_pth_res, file = "Data_file_S2_Differentially_enriched_pathways.csv")

```


### Data file S3
03/31/21
```{r}

load("data_raw/gene_signatures_03-18-21.RData")



gene_sig <- 
  unweighted_gene_sig 
gene_sig$sig_idx <- paste0("S", gene_sig$sig_idx)
gene_sig$fens_category <- paste0("C", gene_sig$fens_category)
gene_sig$coefficient <-  factor(gene_sig$coefficient)
levels(gene_sig$coefficient) <-  c("PLI vs. N&S","PF vs. N&S",  "PF vs. PLI")
names(gene_sig)[1:3] <- c("gene_sig_idx","comparison","nafld_pathway_category")


xx <- as.data.frame(org.Hs.eg.db::org.Hs.egSYMBOL) %>%
  setNames(., c("entrezgene", "gene_name"))

upreg_sig <- gene_sig %>% select(-down_genes_entrez) %>% separate_rows(up_genes_entrez, sep=";")
names(upreg_sig)[4] <- "entrezgene"
upreg_sig <- left_join(upreg_sig, xx) %>% 
  group_by(sig_idx, comparison, nafld_pathway_category) %>% 
  summarise(`up-regulated_gene_names`=paste(unique(gene_name), collapse = ";"),
            `up-regulated_entrez_ids`=paste(unique(entrezgene), collapse = ";"))

downreg_sig <- gene_sig %>% select(-up_genes_entrez) %>% separate_rows(down_genes_entrez, sep=";")
names(downreg_sig)[4] <- "entrezgene"
downreg_sig <- left_join(downreg_sig, xx) %>% 
  group_by(sig_idx, comparison, nafld_pathway_category) %>% 
  summarise(`down-regulated_gene_names`=paste(unique(gene_name), collapse = ";"),
            `down-regulated_entrez_ids`=paste(unique(entrezgene), collapse = ";"))

out_df <- left_join(upreg_sig, downreg_sig)
# write_csv(out_df, "Data_file_S3_Gene_signatures.csv")

```



### Data file S4

```{r}

gene_sig_map <- 
  read_csv("Data_file_S3_Gene_signatures.csv") %>% 
  select(1:3)


cmap_res <- read_rds("unweighted_res_sub_w_p-values_03-22-21.rds")
# Fixing some issues with p-values, since I didn't divide by 
# length. 
cmap_res <-  cmap_res %>% group_by(sig_idx) %>% mutate(p_value=p_value/max(p_value))
cmap_res <-  cmap_res %>% group_by(sig_idx) %>% mutate(fdr_p_value=p.adjust(p_value, method = 'fdr'))
names(cmap_res)[1] <- 'gene_sig_idx'
cmap_res <- left_join(gene_sig_map, cmap_res)

drg_tgts <- read_rds("data_raw/drugbank_db_metadata_v5.1.4.rds") %>% select(drugbank_id, targets) %>% distinct %>% na.omit


trt_cp <-
  read_rds('../cmapAnalysis/GSE92742_inst_metadata.rds') %>% 
  filter(pert_type == "trt_cp") %>% 
  select(sig_id, pert_id, pert_iname) %>% 
  distinct 

annots <- read_rds("data_raw/GSE92742_perts_in_DrugBank_03-30-21.rds")
trt_cp <- inner_join(trt_cp, annots)

out_res <- cmap_res %>% filter(abs(cmap_score) > 0)
out_res <- inner_join(trt_cp, out_res) %>%  relocate(gene_sig_idx, comparison, nafld_pathway_category)
out_res <- out_res %>% 
  group_by(gene_sig_idx) %>% 
  arrange(cmap_score, .by_group=TRUE) %>% 
  mutate(sig_rank=row_number()) %>% 
  ungroup
out_res <- left_join(out_res, drg_tgts)
out_res <- out_res %>% select(-es_up, -es_down, -sig_rank)
out_res <- out_res %>% relocate(targets, .after=name)
# write_csv(out_res, "Data_file_S4_All_drugs.csv")

```

### Data file S5
```{r}

fdr05_tbl <- read_rds('fdr05_tbl_top20_drugs.rds')

names(fdr05_tbl)[4] <- "gene_sig_idx"
cols_to_remove <- c("es_up","es_down" ,"sig_idx","fixed_drug_names"   )
out_file <- fdr05_tbl %>% select(-one_of(cols_to_remove))
out_file <- out_file %>% 
  relocate(gene_sig_idx) %>%  
  relocate(drugbank_id, name, .after=pert_iname) %>% 
  relocate(nafld_pathway_category, .after=gene_sig_idx) %>% 
  relocate(drug_idx, .after=name)
names(out_file)[8] <- "drug_rank"
out_file <- out_file %>% relocate(comparison, .after=gene_sig_idx)
# write_csv(out_file, file = "Data_file_S5_CMap_prioritized_139_compounds.csv")

```



### Data file S6
03/31/21
```{r}

#TODO Need to add in the python code and 

drug_list <- read_rds("manual_lincs_drugbank_drug_annots.rds")
drug_list$drug_name <- drug_list$pert_iname
drug_list$drug_name[str_detect(drug_list$pert_iname, "BR")] <- 
  drug_list$name[str_detect(drug_list$pert_iname, "BR")]

network_res <- 
  read_csv("../network_proximity/network_proximity_results_03-27-21.csv") %>% 
  arrange(z)
names(network_res)[1] <- "drug_name"
network_res <- network_res %>% relocate(z, .after=drug_name)

tmp_drg_ids <- drug_list %>% select(drugbank_id, drug_name) %>% distinct


network_res <- left_join(network_res, tmp_drg_ids)
network_res <- network_res %>% relocate(drugbank_id, .after=drug_name)

# write_csv(network_res, file="Data_file_S6_network_proximity_results.csv")

```
